matrix/src/lib.rs

```diff
@@ -174,25 +177,42 @@ pub trait Matrix<T: Send + Sync>: Send + Sync {
     }
 
     /// Compute Mᵀv, aka premultiply this matrix by the given vector,
-    /// aka scale each row by the corresponding entry in `v` and take the row-wise sum.
+    /// aka scale each row by the corresponding entry in `v` and take the sum across rows.
     /// `v` can be a vector of extension elements.
     #[instrument(level = "debug", skip_all, fields(dims = %self.dimensions()))]
     fn columnwise_dot_product<EF>(&self, v: &[EF]) -> Vec<EF>
     where
         T: Field,
         EF: ExtensionField<T>,
     {
-        self.par_rows().zip(v).par_fold_reduce(
-            || EF::zero_vec(self.width()),
-            |mut acc, (row, &scale)| {
-                acc.iter_mut().zip(row).for_each(|(a, x)| *a += scale * x);
-                acc
-            },
-            |mut acc_l, acc_r| {
-                acc_l.iter_mut().zip(acc_r).for_each(|(l, r)| *l += r);
-                acc_l
-            },
-        )
+        let packed_width = self.width().next_multiple_of(T::Packing::WIDTH);
+
+        let packed_result = self
+            .par_padded_horizontally_packed_rows::<T::Packing>()
+            .zip(v)
+            .par_fold_reduce(
+                || vec![EF::ExtensionPacking::zero(); packed_width],
```

Comment:
Nit: Not sure how expensive this allocation is, but it might be worth trying to use the new `Field::zero_vec()` function (adapted to accommodate packing). Not necessary for this PR though.

Reply:
Yeah it's a good idea, just need to generalize it - I guess it can either be moved from `Field` to `AbstractField`, or perhaps a separate `ZeroVec` trait if we end up using it with an even broader set of types like primitive ints.